[Sema] Add memberwise initializers to nominal type declarations.

dan-zheng opened the pull request:
Always add synthesized memberwise initializers to nominal type declarations
instead of conformance contexts, which may be extensions. This is safe
because derived conformances cannot be requested from cross-file extensions
(see https://github.com/apple/swift/pull/24538).

Resolves [TF-574](https://bugs.swift.org/browse/TF-574).

dan-zheng commented:
@swift-ci Please test tensorflow

jrose-apple commented:
I know this is the TensorFlow branch, but that's not going to be correct. If you don't have access to the original StructDecl, you don't get to synthesize a memberwise init.

dan-zheng commented:
> I know this is the TensorFlow branch, but that's not going to be correct. If you don't have access to the original StructDecl, you don't get to synthesize a memberwise init.

I see - I wonder if this aspect of memberwise initializer semantics is documented somewhere?

In this case, I think adding memberwise initializers to the original type declaration instead of the conformance context may also work. Let me try it.

jrose-apple commented:
I mean, the mainline compiler doesn't have a notion of "memberwise initializer" because "something that can be synthesized". It's more about who's allowed to request that something be synthesized. If there are `private` properties, for example, it wouldn't be okay to request a memberwise initializer in another file, because that would expose them.

dan-zheng committed:
[Sema] Add memberwise initializers to nominal type declarations.

Always add synthesized memberwise initializers to nominal type declarations
instead of conformance contexts, which may be extensions. This is safe
because derived conformances cannot be requested from cross-file extensions.

Resolves TF-574.

dan-zheng commented:
> I mean, the mainline compiler doesn't have a notion of "memberwise initializer" because "something that can be synthesized". It's more about who's allowed to request that something be synthesized. If there are `private` properties, for example, it wouldn't be okay to request a memberwise initializer in another file, because that would expose them.

I see, avoiding `parentDC` as an unsafe customization point for `swift::createImplicitConstructor` is definitely desirable.

I changed derived conformances logic to always add memberwise initializers to nominal type declarations instead of conformance contexts. I believe tests continue to pass.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
